packages/shared/src/UI/components/AssetPlayer/index.tsx

```diff
@@ -51,53 +49,107 @@ enum AssetPlayerState {
+    )
 
-    // #region If onResized is not triggered within the specified time, set player state to error
-    const [, cancel, reset] = useTimeoutFn(() => {
-        if (playerState !== AssetPlayerState.NORMAL) {
-            setPlayerState(AssetPlayerState.ERROR)
-        }
-    }, TIMEOUT)
-    // #endregion
+    useEffect(() => {
+        if (!props.renderTimeout || !hidden) return
+
+        const timer = setTimeout(() => {
+            setHidden(false)
+        }, props.renderTimeout)
+
+        return () => clearTimeout(timer)
+    }, [props.renderTimeout, hidden])
 
     // #region setup iframe when url and options be changed
     const setIframe = useCallback(() => {
         // if iframe isn't be init or the load error has been existed
-        if (!ref.current || playerState === AssetPlayerState.ERROR) return
-        if (!url) {
+        if (!ref.current || playerState === AssetPlayerState.ERROR || playerState === AssetPlayerState.NORMAL) return
+        if (!url && !erc721Token) {
             setPlayerState(AssetPlayerState.ERROR)
             return
         }
-        if (playerState === AssetPlayerState.INIT || playerState === AssetPlayerState.NORMAL) {
-            reset()
+        if (playerState === AssetPlayerState.INIT) {
             ref.current.iFrameResizer.sendMessage({
                 url,
+                erc721Token,
                 type,
                 ...options,
             })
             return
         }
-    }, [url, type, JSON.stringify(options), playerState])
-    // #endregion
+    }, [url, JSON.stringify(erc721Token), type, JSON.stringify(options), playerState])
+    // endregion
 
+    type ERC721TokenNameMsg = { message: { type: 'name'; name: string } | { type: 'sourceType'; name: string } }
     // #region resource loaded error
-    const onMessage = useCallback(({ message }: { message: { name: string } }) => {
-        if (message?.name === 'Error') {
-            setPlayerState(AssetPlayerState.ERROR)
-        }
-    }, [])
+    const onMessage = useCallback(
+        ({
+            message,
+        }: {
+            message: { name: string } | ERC721TokenNameMsg | { type: 'webglContextLost' } | { type: 'reload' }
+        }) => {
+            if ((message as { name: string })?.name === 'Error') {
+                setPlayerState(AssetPlayerState.ERROR)
+            }
+            if ((message as { type: 'webglContextLost' })?.type === 'webglContextLost') {
+                setHidden(true)
+                setPlayerState(AssetPlayerState.LOADING)
+                setTimeout(() => setHidden(false), 1000)
+            }
+            if ((message as ERC721TokenNameMsg).message?.type === 'name') {
+                props.setERC721TokenName?.((message as ERC721TokenNameMsg).message.name)
+            }
+            if ((message as ERC721TokenNameMsg).message?.type === 'sourceType') {
+                props.setSourceType?.((message as ERC721TokenNameMsg).message.name)
+            }
```

Comment:
@Jack-Works approve?

Reply:
We can omit `as` operator with https://www.typescriptlang.org/docs/handbook/release-notes/overview.html#type-guards-inferred-from-in-operator